pkg/kubelet/cm/qos_container_manager_linux.go

```diff
@@ -0,0 +1,155 @@
+	nodeInfo          *v1.Node
+	qosContainersInfo QOSContainersInfo
+	subsystems        *CgroupSubsystems
+	cgroupManager     CgroupManager
+	activePods        ActivePodsFunc
+}
+
+func NewQOSContainerManager(subsystems *CgroupSubsystems, nodeConfig NodeConfig) (QOSContainerManager, error) {
+	if !nodeConfig.CgroupsPerQOS {
+		return &qosContainerManagerNoop{
+			cgroupRoot: CgroupName(nodeConfig.CgroupRoot),
+		}, nil
+	}
+
+	// this does default to / when enabled, but this tests against regressions.
+	if nodeConfig.CgroupRoot == "" {
+		return nil, fmt.Errorf("invalid configuration: cgroups-per-qos was specified and cgroup-root was not specified. To enable the QoS cgroup hierarchy you need to specify a valid cgroup-root")
+	}
+
+	// we need to check that the cgroup root actually exists for each subsystem
+	// of note, we always use the cgroupfs driver when performing this check since
+	// the input is provided in that format.
+	// this is important because we do not want any name conversion to occur.
+	cgroupManager := NewCgroupManager(subsystems, "cgroupfs")
+	if !cgroupManager.Exists(CgroupName(nodeConfig.CgroupRoot)) {
+		return nil, fmt.Errorf("invalid configuration: cgroup-root doesn't exist")
+	}
+	glog.Infof("container manager verified cgroup-root exists: %v", nodeConfig.CgroupRoot)
+
+	return &qosContainerManagerImpl{
+		NodeConfig:    nodeConfig,
+		subsystems:    subsystems,
+		cgroupManager: NewCgroupManager(subsystems, nodeConfig.CgroupDriver),
+	}, nil
+}
+
+func (m *qosContainerManagerImpl) GetQOSContainersInfo() QOSContainersInfo {
+	return m.qosContainersInfo
+}
+
+func (m *qosContainerManagerImpl) Start(nodeInfo *v1.Node, activePods ActivePodsFunc) error {
+	// Top level for Qos containers are created only for Burstable
+	// and Best Effort classes
+	qosClasses := [2]v1.PodQOSClass{v1.PodQOSBurstable, v1.PodQOSBestEffort}
+
+	// Create containers for both qos classes
+	for _, qosClass := range qosClasses {
+		// get the container's absolute name
+		absoluteContainerName := CgroupName(path.Join(m.CgroupRoot, string(qosClass)))
+		// containerConfig object stores the cgroup specifications
+		containerConfig := &CgroupConfig{
+			Name:               absoluteContainerName,
+			ResourceParameters: &ResourceConfig{},
+		}
+		// check if it exists
+		if !m.cgroupManager.Exists(absoluteContainerName) {
+			if err := m.cgroupManager.Create(containerConfig); err != nil {
+				return fmt.Errorf("failed to create top level %v QOS cgroup : %v", qosClass, err)
+			}
+		}
+	}
+
+	// Store the top level qos container names
+	m.qosContainersInfo = QOSContainersInfo{
+		Guaranteed: m.CgroupRoot,
+		Burstable:  path.Join(m.CgroupRoot, string(v1.PodQOSBurstable)),
+		BestEffort: path.Join(m.CgroupRoot, string(v1.PodQOSBestEffort)),
+	}
+	m.nodeInfo = nodeInfo
+	m.activePods = activePods
+
+	return nil
+}
+
+func (m *qosContainerManagerImpl) UpdateCgroups() error {
+	m.Lock()
+	defer m.Unlock()
+
+	qosConfigs := map[v1.PodQOSClass]*CgroupConfig{
```

Comment:
do you mind if we add this in a follow-on and just have this return nil after the lock/defer lock call?